rust/src/rdp/parser.rs

```diff
@@ -0,0 +1,1559 @@
+            },
+            Err(e) => return Err(e),
+        },
+    };
+    let child: T123TpktChild = match opt3 {
+        Some(x) => x,
+        None => T123TpktChild::Raw(data.to_vec()),
+    };
+
+    return Ok((i4, T123Tpkt { child }));
+}
+
+/// rdp-spec, section 2.2.1.1
+pub fn parse_x224_connection_request(
+    input: &[u8],
+) -> IResult<&[u8], X224ConnectionRequest> {
+    let (i1, length) = try_parse!(input, verify!(be_u8, |x| x != 0xff)); // 0xff is reserved
+    let (i2, cr_cdt) = try_parse!(
+        i1,
+        bits!(tuple!(
+            verify!(take_bits!(u8, 4), |x| x
+                == X224Type::ConnectionRequest as u8),
+            verify!(take_bits!(u8, 4), |x| x == 0 || x == 1)
+        ))
+    );
+    let (i3, dst_ref) = try_parse!(i2, verify!(be_u16, |x| x == 0));
+    let (i4, src_ref) = try_parse!(i3, be_u16);
+    let (i5, class_options) = try_parse!(
+        i4,
+        bits!(tuple!(
+            verify!(take_bits!(u8, 4), |x| x <= 4),
+            verify!(take_bits!(u8, 4), |x| x <= 3)
+        ))
+    );
+    // less cr_cdt (u8), dst_ref (u16), src_ref (u16), class_options (u8)
+    let (i6, sz) = try_parse!(i5, expr_opt!(length.checked_sub(6)));
+
+    //
+    // optionally find cookie and/or negotiation request
+    //
+
+    let (i7, data) = {
+        if sz > 0 {
+            try_parse!(i6, take!(sz))
+        } else {
+            (i6, &[][..])
+        }
+    };
+
+    let (j1, cookie) = {
+        if data.len() > 0 {
+            match opt!(data, parse_rdp_cookie) {
+                Ok((remainder, opt)) => (remainder, opt),
+                Err(e) => return Err(e),
+            }
+        } else {
+            (&[][..], None)
+        }
+    };
+
+    let (j2, negotiation_request) = {
+        if j1.len() > 0 {
+            match opt!(j1, parse_negotiation_request) {
+                Ok((remainder, opt)) => (remainder, opt),
+                Err(e) => return Err(e),
+            }
+        } else {
+            (&[][..], None)
+        }
+    };
+
+    return Ok((
+        i7,
+        X224ConnectionRequest {
+            cdt: cr_cdt.1,
+            dst_ref,
+            src_ref,
+            class: class_options.0,
+            options: class_options.1,
+            cookie,
+            negotiation_request,
+            data: j2.to_vec(),
+        },
+    ));
+}
+
+/// rdp-spec, section 2.2.1.1
+/// "An X.224 Class 0 Connection Request TPDU, as specified in [X224] section 13.3."
+pub fn parse_x224_connection_request_class_0(
+    input: &[u8],
+) -> IResult<&[u8], X224ConnectionRequest> {
+    let (i1, x224) = try_parse!(input, parse_x224_connection_request);
+    if x224.class == 0 && x224.options == 0 {
+        Ok((i1, x224))
+    } else {
+        Err(nom::Err::Error(error_position!(
+            input,
+            ErrorKind::Custom(RDP_NOT_X224_CLASS_0_ERROR)
+        )))
+    }
+}
+
+// rdp-spec, section 2.2.1.1.1
+named!(pub parse_rdp_cookie<RdpCookie>,
+    do_parse!(
```

Comment:
Info: rustfmt tends to mess up indentation in `do_parse` macros. However, if you use '{' instead of '(', it will not, and the macro can be nicely indented:
```rust
do_parse! {
  _key: verify!(...) >>
  _name: verify!(...) >>
```

Also (minor), `"Cookie: ".as_bytes()` can be written `b"Cookie: "`

Reply:
Good tip for `do_parse` and rustfmt!  Will incorporate both suggestions.